ssz_serialization/merkleization.nim

```diff
@@ -120,23 +120,82 @@ func digest*(a: openArray[byte]): Digest {.noinit.} =
         h.update(a)
         h.finish()
 
-func digest(a, b: openArray[byte]): Digest =
-  result = computeDigest:
-    trs "DIGESTING ARRAYS ", toHex(a), " ", toHex(b)
-    trs toHex(a)
-    trs toHex(b)
-
-    h.update a
-    h.update b
-  trs "HASH RESULT ", result
+{.compile: "hashtree.c".}
+
+when defined(gcc):
+  when defined(arm):
+    {.compile: "../vendor/hashtree/src/sha256_armv8_neon_x1.S".}
+    {.compile: "../vendor/hashtree/src/sha256_armv8_neon_x4.S".}
+
+  elif defined(amd64):
+    {.compile: "../vendor/hashtree/src/sha256_avx_x1.S".}
+    {.compile: "../vendor/hashtree/src/sha256_avx_x4.S".}
+    {.compile: "../vendor/hashtree/src/sha256_avx_x8.S".}
+    {.compile: "../vendor/hashtree/src/sha256_avx_x16.S".}
+    {.compile: "../vendor/hashtree/src/sha256_shani.S".}
+    {.compile: "../vendor/hashtree/src/sha256_sse_x1.S".}
+
+type Hasher = proc(output: pointer, input: pointer, count: uint64) {.cdecl, noSideEffect, raises: [Defect].}
+
+func digest64x(output: pointer, input: pointer, count: uint64) {.cdecl, noSideEffect, raises: [Defect].} =
+  cast[ptr Digest](output)[] =
+     digest(cast[ptr UncheckedArray[byte]](input).toOpenArray(0, 63))
+  discard
+
+func best_sha256_implementation(): Hasher {.importc.}
+
+let digest64* = block:
+  let hasher = best_sha256_implementation()
+  if isNil(hasher):
+    (Hasher)digest64x
+  else:
+    hasher
 
-func digest(a, b, c: openArray[byte]): Digest =
-  result = computeDigest:
-    trs "DIGESTING ARRAYS ", toHex(a), " ", toHex(b), " ", toHex(c)
+func digest(a, b: openArray[byte]): Digest {.noinit.} =
+  when true:
+    result = computeDigest:
+      h.update a
+      h.update b
+  else:
+    if a.len + b.len == 64:
+      {.noSideEffect.}:
+        if a.len == 64:
+          digest64(addr result.data[0], unsafeAddr a[0], 1)
+        else:
+          var tmp {.noinit.}: array[64, byte]
+          copyMem(addr tmp[0], unsafeAddr a[0], a.len)
+          copyMem(addr tmp[a.len], unsafeAddr b[0], b.len)
+          digest64(addr result.data[0], addr tmp[0], 1)
+    else:
+      result = computeDigest:
+        h.update a
+        h.update b
+  trs "HASH RESULT ", result
 
-    h.update a
-    h.update b
-    h.update c
+func digest(a, b, c: openArray[byte]): Digest {.noinit.} =
```

Comment:
When does this happen?

Reply:
when joining a 32-byte data chunk with a <32-byte chunk - `c` is the zero padding mandated by ssz